doc/design/listener.md

```diff
@@ -0,0 +1,306 @@
+a separate background thread in which the event-callbacks are executed in contrast
+to the WaitSet where the user has to call the event-callbacks explicitly.
+
+## Terminology
+
+- [Condition Variable](https://en.wikipedia.org/wiki/Monitor_(synchronization)#Condition_variables_2)
+  Used by an attached object to inform the Listener/WaitSet that an event has
+  occurred.
+- **event** is changing the state of an object.
+- **event driven** a one time reaction which is caused directly by an event.
+      Example: a new sample has been delivered to a subscriber.
+- [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug)
+- [Reactor pattern](https://en.wikipedia.org/wiki/Reactor_pattern)
+  The Listener and WaitSet are both variations of the that pattern.
+- **state** predefined values to which the members of an object are set.
+- **state driven** a repeating reaction which is continued as long as the state
+      persists.
+      Example: a subscriber has stored samples which were not inspected by the user.
+
+## Design
+
+The usage should be similar to the _WaitSet_ with a key difference - it should
+be **event driven** and not a mixture of event and state driven, depending on
+which event is attached, like in the _WaitSet_.
+
+### Requirements
+
+- Whenever an event occurs the corresponding callback should be called **once**
+    as soon as possible.
+- If an event occurs multiple times before the callback was called, the callback
+    should be called **once**.
+- If an event occurs while the callback is being executed the callback should be
+    called again **once**.
+- Thread-safety: attaching an event at any time from any thread.
+- Thread-safety: detaching an event at any time from any thread.
+  - If the callback is currently running `detachEvent` blocks until the callback
+      is finished.
+  - After the `detachEvent` call the event-callback is not called anymore, even
+      when the event was signalled while `detachEvent` was running and the
+      callback was not yet executed.
+  - If the callback is detached from within the event-callback then `detachEvent`
+     is non blocking. The event is detached right after the `detachEvent` call.
+- One can attach at most one callback to a specific event of a specific object.
+  - Usually defined with an enum by the developer. One example is `SubscriberEvent::HAS_SAMPLE_RECEIVED`.
+  - Attaching a callback to an event where a callback has been already attached
+      results in an error.
+- One can attach the same event to different objects at the same time.
+- One can attach multiple different events to a single object
+- When the Listener goes out of scope it detaches itself from every attached
+    object via a callback provided by the attached object (like in the WaitSet).
+- When the class which is attached to the Listener goes out of scope it
+    detaches itself from the Listener via a callback provided by the Listener
+    (like in the WaitSet).
+
+### Solution
+
+#### Class diagram
```

Comment:
Shouldn't we use plantuml for this? I know that it's not that nice to read in plain text, but it could be rendered to a SVG and included as image in the markdown file. What do you think?

Reply:
As I can see there is also the possibility to integrate it into github.
@dkroenke what do you think?

I think plantuml is not so nice to read in a markdown file but you can create a diagram much faster and the svg version is then a nice end result. I am for it!

Reply:
@elfenpiff @elBoberido 
I played around with plantuml on github markdown here: https://github.com/ApexAI/iceoryx/tree/plantuml-github-example/doc/design

The corresponding puml file is stored here: https://github.com/ApexAI/iceoryx/blob/plantuml-github-example/doc/design/puml/puml_github_workflow.puml
And the link to the puml file in markdown is done with that command:
```
![uncached image](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/ApexAI/iceoryx/plantuml-github-example/doc/design/puml/puml_github_workflow.puml)
```
Please note that the link to the puml file needs to use `https://raw.githubusercontent.com`. In this example it points to a feature branch, but later it needs to point to the file on master branch.

Reply:
@dkroenke what do you suggest as workflow?
- puml and markdown in the same PR
- separate PRs, first puml then markdown

Reply:
For the first PRs it make more sense to merge first the puml and afterwards the markdown in the link to check if everything works correct.
After that it would be also fine to merge puml and markdown together with a modified link which points to `eclipse-iceoryx/iceoryx` on master branch in the markdown link. The link then get resolved after merge.

Please note that all links should point to master branch and not any feature branch. Maybe we can operate same with release branches to avoid that text and diagram get out of sync.

Reply:
@elBoberido another option would be to use svg files directly as source of truth and embedd them in markdown: https://github.com/ApexAI/iceoryx/blob/plantuml-github-example/doc/design/puml/svg_plantuml_github%20workflow.svg
See again https://github.com/ApexAI/iceoryx/tree/plantuml-github-example/doc/design for an example.

One thing here was that i needed to set `skinparam monochrome reverse` because the svg background is rendered transparent which is not cool when using dark mode like i do.
This has the advantage that you have one source of truth and svg files are versioned in GH.
I would propose to store the puml file and the exported svg file in the puml folder and link against the svg in markdown.

Reply:
we also have to take care that always the same plantuml version is used, else the problem with huge diffs is even bigger